### PR TITLE
fixed bug that would require customer to click on use new payment met…

### LINKED ISSUE #2163: UPE pay for order page displays "Use a new payment method" radio even if there are no other payment methods saved
**Describe the bug**
UPE pay for order page displays "Use a new payment method" radio button even if there are no other payment methods saved.

If the radio is not selected, this causes an error when trying to pay:

> Payment processing failed. Please retry.

**To Reproduce**
1. Make sure you have no payment methods saved to your account.
2. Add any product to cart.
3. Go to checkout.
4. Add this credit card `4000000000000002`.
5. Checkout.
6. An error message should be displayed.
7. Go to My account > Order > Last order > Pay.
8. Notice the radio button is present and is not automatically selected.

**Expected behavior**
It shouldn't display the radio button when there are no other payment methods saved. And it should have one of the radio buttons automatically selected, the same way as in the checkout page.

**Screenshots**
Stripe pay for order page UPE - use a new payment method

**Additional context**
p1637159154256000-slack-C01BZUL57SQ

### PR DESCRIPTION
**Pull request by leonardola**
Fixes #2163 

## Changes proposed in this Pull Request:
Now the "Use new payment method" radio button only shows on pay order page when there are saved cards. Otherwise it is not needed

Stripe pay for order page UPE - use a new payment method

## Testing instructions
- As a shopper
- With UPE active
- Remove any saved card in My account > Payment Methods
- Buy any product
- Go to checkout
- Select Stripe as the payment gateway
- Make sure the radio button "Use new payment method" does not show up
- Use a **valid** card `4242 4242 4242 4242`
- Click on save credit card
- Place order
- Must be successful
- Repeat but this time use the saved card to checkout
- The saved card radio button and the "Use new payment method" must show 
- Place order
- Must be successful

- Remove any saved card in My account > Payment Methods
- Buy any product
- Go to checkout
- Select Stripe as the payment gateway
- Make sure the radio button "Use new payment method" does not show up
- Use an **invalid** card `4000000000000002`
- Try to place order
- Go to My account > Orders
- Click on pay on the last order
- Select Stripe
- Make sure the radio button "Use new payment method" does not show up
- Select Stripe as the payment method
- Use a valid card `4242 4242 4242 4242`
- Place the order
- Must be successful
- Repeat but this time use the saved card to checkout on pay order page
- The saved card radio button and the "Use new payment method" must show 
- Place order
- Must be successful

---

-   [x] Added changelog entry (or does not apply)

**Post merge**

-   [x] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

### CODE REVIEW

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 1]
```diff
@@ -438,7 +438,9 @@ public function payment_fields() {
 			<?php
 			if ( $display_tokenization ) {
 				$this->tokenization_script();
-				$this->saved_payment_methods();
+				if ( $this->get_tokens() ) {
```
Comment (a-danae) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: This works well. My doubt is why, without this change, things work well on the regular checkout but not on the Pay for order page?

This seems to be handled on the client-side on the checkout page. It also works well in both the Checkout and Pay for order pages in WCPay and we don't check for `$this->get_tokens()` before calling `$this->saved_payment_methods()` there. Again, this works. I'm wondering if there's a small detail we're missing here for things to work as expected.

**Review by a-danae — COMMENTED** [review]
Hey!

This works well and makes sense to me. I left a question regarding other approaches we _might_ check out. Thanks for working on this, @leonardola!

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php (hunk 1 above)
Comment (leonardola) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: During checkout the payment methods element gets re-rendered right after the page is loaded which ends up removing the options when they are not necessary. I could not find any JS that removes the options on the plugin apparently thats something that is done by WooCommerce itself. I'll take a deeper look to see if there's anything there.

I have written another solution with js only which would wait till everything loads and check if the options are necessary and hide it when not. But to me it makes more sense to not render it at all when not necessary

Comment (leonardola) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php]: I found that [this file](https://github.com/woocommerce/woocommerce/blob/release/5.9/assets/js/frontend/tokenization-form.js#L109) handles the form on checkout and it is triggered [here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/classic/upe.js#L288) on WC Pay

So I copied that to Stripe plugin and it works too. Can you please take another look at this?


Resolved with 222fbdee73dfa62bc70b50d2b91628d19d251993

**Comment by a-danae** [comment]
Sweet, things work fine and this is also how we're handling it in WC Pay. Thanks for giving it a second look, @leonardola! LGTM 👍 

(We'd need to move the changelog entry to the latest version, btw)

**Review by a-danae — APPROVED** [review]
Sweet, things work fine and this is also how we're handling it in WC Pay. Thanks for giving it a second look, @leonardola! The JS linting and unit tests failures don't seem to be related to this change, so LGTM 👍

(We'd need to move the changelog entry to the latest version, btw)

**Pull request merged**